### PR TITLE
DB + authentication

### PR DESCRIPTION
**Pull request by cjdenio**
This lovely PR adds Slack-powered authentication (as well as a Diesel-powered database), _mostly_ compatible with the previous iteration of HaaS's API. Notable differences include:
- Use of Slack's new OpenID Connect API (instead of plain OAuth)
- Lowercase field names
- All routes are under `/api`, where previously the login route was `/oauth/login`

### CODE REVIEW

File: src/auth.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    http::Status,
+    request::{self, FromRequest, Outcome, Request},
+};
+
+use crate::{
+    models::{token::Token, user::User},
+    DbConn,
+};
+
+pub struct HaaSUser(pub User);
```
Comment (anirudhb) [inline src/auth.rs]:
```suggestion
pub struct HaasUser(pub User);
```

Follow rust style conventions

**Review by anirudhb — APPROVED** [review]
Looks great to me! Two things:

* Please open a frontend pr for the incompatible changes (and update the swagger file accordingly)
* If it's possible to add SQLite support in this pr (pluggable db backends) I would appreciate that (add an SQLite feature that writes to a file instead of using Postgres)

**Comment by cjdenio** [comment]
I don't think multi-database support is feasible because migrations are written in SQL and Postgres/SQLite may have tiny inconsistencies with types and such (it works for Rails because migrations are written in a database-agnostic DSL)

**Comment by cjdenio** [comment]
Maybe we should move the Swagger definition to this repo instead of hack-as-a-service/frontend?

**Pull request merged**